CCDirector Refactoring and un-singletoning

andykorth opened the pull request:
CCDirector Refactoring:

The primary goal behind this pull request is to un-singleton the CCDirector. This allows for better scene control, better testability, and clearer separation of platform specific code. CCViews own the reference to the CCDirector, and CCDirectors manage and own individual CCScenes.

Most of the credit goes to Oleg and Scott. Thanks!

Sorry, this is a huge pull request. I was initially breaking it down into smaller commits, but found each one depended on the previous. This pull request is going out now to prevent even larger ones from happening in the future. While this refactoring is not entirely complete, the project is currently working and usable. iOS unit tests, iOS testbed, and Mac has been recently tested. Oleg is working on Android.

Highlights of changes
- CCDirector is no longer a singleton, but the current director is accessible, similar to how CCRenderer is accessible.
- CCView is a common protocol to each platforms view implementation.
- CCScenes have a clearer relationship with directors.
- CCActions were refactored to reflect the possibility that they were added to nodes that haven't yet been added to a scene
  *\* CCActionManager is removed and merged into CCScheduler. There was a lot of common functionality there without any common code.
  *\* Merging these classes unified pausing behavior. 
  *\* Pretty big performance improvements both for actions and scheduled timers. (calling update on nodes, etc)
- Refactor scene changing methods. Use presentScene exclusively to change scenes, whether it's the first scene or not.

Areas of continued development:
- Running CCActions in the fixed update loop.
- Queuing scheduled timers on nodes that haven't been added to a scene
- Fixing up test like the package manager tests (testability without a view, etc)
- ContentScale moving out of director.
- Stats label
- Application lifecycle changes w/ director.

Ultimately, there are too many people working on highly related code for these changes not to live in a commonly accessible place. Even if these changes aren't super stable, the alternative is people writing large amount of code independently and having huge merge conflicts at the end. I'm not sure what exactly the stability expectations are of the "develop" branch, but that would come at the cost of regular integration and conflicts.

krazyeom committed:
Conflicts:
	external/ObjectAL

andykorth committed:
Prevent CCScenes from being added to other scenes

andykorth committed:
Merge branch 'develop-v4' of github.com:cocos2d/cocos2d-swift into CCDirectorUpdates

andykorth committed:
Extracting scheduler and actionManager from CCDirector

andykorth committed:
Tracking down action failure in a clip node.

andykorth committed:
Fix stencil node and several tests, parenting behavior.

osinoleg committed:
Adding support for multiple CCViews
Pass 1 - Replaced CCViewGL for the iOS and Mac targets.

Removed CCGLView from CCTouch and CCResponderManager.

andykorth committed:
Accessing views and directors from CCNode.

andykorth committed:
CCScene's reference to director should be weak

osinoleg committed:
Replaced the use of [CCDirector sharedDirector] wherever possible, with the use of self.director.

andykorth committed:
Separate CCScheduler into public and private headers

andykorth committed:
Simplify fixed mode of CCActionManager in new CCDirector.

andykorth committed:
Merge and add test comments.

andykorth committed:
merge

andykorth committed:
Merge branch 'CCDirectorRefactorActionManagerChanges' into CCDirectorRefactorSchedulerChanges

andykorth committed:
Merge branch 'develop-v4' of github.com:cocos2d/cocos2d-swift into CCDirectorRefactorSchedulerChanges

Conflicts:
	cocos2d.xcodeproj/project.pbxproj
	cocos2d/Platforms/Android/CCGLView.h

osinoleg committed:
Added a currentDirector method that uses thread local storage, replacement for the singleton patter for sharedDirector.

osinoleg committed:
Changed CCDirector access inside CCAction.
Removed the use of self.director in CCNode onEnter - because it's possible that onEnter is called before a CCNode has a parent (look at CCRendererTests:setupClippingNodeTest).

andykorth committed:
Remove unneeded method from private header

andykorth committed:
Add new CCAction tests, including a performance test.

andykorth committed:
Rename SchedulerTarget to SchedulableTarget to reduce confusion with ScheduledTarget

andykorth committed:
Initial working commit of merged action and scheduler.

andykorth committed:
Remove old ActionManager references. Ensure actions can pause

andykorth committed:
merge

andykorth committed:
CCScheduledTarget work

andykorth committed:
copy on write arrays for scheduler

andykorth committed:
profiling, performance adjustments, fix removals

andykorth committed:
CCScheduler is noarc

slembcke committed:
CCScheduler cleanup and some fixes.

slembcke committed:
CopyOnWriteArray comments and fixes.

slembcke committed:
More scheduled target refactoring

andykorth committed:
Merge pull request #2 from slembcke/MergeSchedulerAndActionManager

Merge scheduler and action manager work.

andykorth committed:
Replaced O(n) removal with constant time remove from target.

andykorth committed:
Remove __bridge from no-arc file

andykorth committed:
Scheduler and action performance tests

andykorth committed:
Respect paused state when adding actions

andykorth committed:
Provide different backing data structures to scheduler's copy on write datastrcutre

andykorth committed:
Update mac ccview

andykorth committed:
Merge branch 'CCDirectorRefactorSchedulerChanges' into MacKeyboard

Conflicts:
	cocos2d.xcodeproj/project.pbxproj
	cocos2d/CCResponderManager.m

andykorth committed:
Quick Mac keyboard support.

andykorth committed:
Merge branch 'MacKeyboard' into CCDirector/MergeSchedulerAndActionManager

Conflicts:
	cocos2d.xcodeproj/project.pbxproj
	cocos2d/CCResponderManager.m

andykorth committed:
Merge remote-tracking branch 'cocos/develop' into CCDirector/MergeSchedulerAndActionManager

andykorth committed:
Merge remote-tracking branch 'oleg/ccview_refactor_v4' into CCDirector/MergeSchedulerAndActionManager

Conflicts:
	cocos2d-ui/CCBReader/CCBReader.m
	cocos2d/CCNode.m
	cocos2d/CCResponderManager.m
	cocos2d/Platforms/Mac/CCGLView.m

andykorth committed:
Missed a conflict.

slembcke committed:
1 director per view.

andykorth committed:
Merge branch 'CCDirectorUpdates' into CCDirector/MergeSchedulerAndActionManager

Conflicts:
	UnitTests/CCNodeTests.m
	UnitTests/CCPhysicsTests.m
	cocos2d-ui/CCBReader/CCAnimationManager.m
	cocos2d-ui/CCBReader/CCBReader.m
	cocos2d.xcodeproj/project.pbxproj
	cocos2d/CCDirector.h
	cocos2d/CCDirector.m
	cocos2d/CCDirector_Private.h
	cocos2d/CCNode.m
	cocos2d/CCScene.h
	cocos2d/CCScene.m

andykorth committed:
Update tests for new scheduler location

andykorth committed:
Merge remote-tracking branch 'scottLocal/CCDirector' into CCDirector/MergeSchedulerAndActionManager

andykorth committed:
Fixes and assertions for failed responder stuff

andykorth committed:
Scene transitions need the right director setup.

andykorth committed:
Tests for action queueing.

andykorth committed:
Queue actions if they can't be added immediately

andykorth committed:
Change how scenes access non-singleton director

andykorth committed:
Plugging leaks

andykorth committed:
Ensure TTF labels work with the new director changes

andykorth committed:
Use presentScene exclusively to change scenes. Refactor duplicate and unused logic in scene changes.

andykorth committed:
Rename start/stopAnimation to start/stopRunLoop, to reflect its actual purpose.

andykorth committed:
Fixing unit tests. Disabling package tests for now

andykorth committed:
Merge branch 'develop' of github.com:cocos2d/cocos2d-swift into CCDirector/MergeSchedulerAndActionManager

andykorth committed:
Fix mac and android build targets

andykorth committed:
Mac director fixes

andykorth committed:
Main loop refactor. Fixes for Mac window/view size

rkachowski commented on cocos2d/Platforms/Mac/CCViewMacGL.m:
```diff
@@ -0,0 +1,296 @@
+- (void)mouseDown:(NSEvent *)theEvent
+{
+    // dispatch mouse to responder manager
+    [CCDirector bindDirector:_director];
```
What is the benefit here? It looks like we are adding 2 lines of boilerplate overhead to every `CCDirector` call

andykorth replied:
Yeah, unfortunately the bindDirector calls are pretty verbose in this section. The reason for this is that each one of these events is an entry point into cocos2d code from various NS input handling places. Further down this stack trace, we will have user code responding to mouseDown events. Inside that user code, the user might be looking for the `[CCDirector currentDirector]` for that thread, and it needs to be bound for them. 

I'm open to other suggestions.

It could be moved into the scheduler, but I think that's just pushing the binds into bigger methods where it's not as easy to see what's going on. (although looking at it, there would be a few less calls).

And it's important to unbind the director to ensure correctness. It makes it much easier to find a mistake like a missing bind in the engine, rather than leaving the 'usually correct' director bound, causing infrequent race-condition like issues when the user's setting up a new scene or something like that.

Thank you for reviewing!

rkachowski replied:
As far as the duplication goes, at the least I would create some kind of method to encapsulate the binding/unbinding of all this. e.g. `[foo withDirector:_director perform:^(){[_director.responderManager mouseDown:theEvent];}]` - also the (duplicated) comments here are quite superfluous, as the code is rather explicit about what it is doing.

On a larger architectural scale, it seems pretty inefficient to forward every event to a secondary object with the exact same interface. There's no benefit of platform abstraction here. Why not have a single `handleUIEvent:` method on the responderManager, which will invoke different logic depending on the event type? as it stands there is a 1:1 mapping between the API of these classes.

Is there a situation where different threads would have different directors? The continuous binding and unbinding feels like a rather major anti-pattern to introduce into the codebase.

andykorth replied:
Good ideas! The extra mirrored methods in CCScheduler are definitely superfluous. I'll update the request. The CCViewMacGL has way more events than iOS, they can definitely be sent to a much smaller number of already existing methods. It could be converted to mouse, scroll wheel, and key events pretty trivially, and could be convered to a single handleUIEvent method too (although organizationally that might be a bit of a big method).

Yeah, I should have removed those nonsense comments when I was touching it. CCViewMacGL is a renamed CCGLView, which is pretty old. Will cleanup.

There are situations where the same thread could have different directors (and different views). This is potentially addressing situations where a game is running with two CCViews, each one would have different directors. I'm guessing we can't make any assumptions over which thread is calling these input methods. Maybe @slembcke can comment more on this. It also improves testability, when we can create fresh CCDirectors for each test.  Any other ideas for improving this? The only one that really comes to mind is "bring back the singleton" or do a better job of encapsulating the bindDirector calls (which I am on!).

slembcke replied:
One of the biggest problems with the director being a singleton is that you can only have one cocos2D view in an application, and it makes the startup process _really_ convoluted. (as I'm sure you've discovered) Since it's not accessible as a global anymore, it either needs to be passed around as an argument or set as a context value. Passing the director as an argument to all update methods, events, user callbacks, etc is also a little crazy and a _lot_ more work for both us and users. Setting it as a context value lets it work in a more or less backwards compatible way for the majority of users that will only ever use a single view in their application.

We don't really have a lot of choice in forwarding the methods I don't think. Either every view type (currently 4 of them) needs to implement it's own event dispatching, or we need to forward the events to a common object that is shared by all of the platforms. Efficiency isn't really a problem. Events happen much less often than the rendering loop, so a few extra method calls aren't going to hurt the performance. It's Basically the same reason why we decided not to remove the director after thinking about it more. The code that is currently in the director would just need to be duplicated by 4 different view classes. On the other hand, we now have the opportunity to merge the 3 director classes into one since there isn't very much x-platform code in it anymore that couldn't move to the view classes.

Since we can't really break the dependence on having a director, or at the very least, knowing the view that a certain scene is running as a part of, we don't really have any other ideas of what to do.

Also, I suppose it's important to note that the mouse\* events are defined on NSResponder so we can't really change how that works.

andykorth replied:
Yeah NSResponder could have been shorter.

Opinions on these selectors? Mash them together? Not sure if I should change something that will break user code, but if there's ever a chance for an API change, it's now. Plus I bet the the mac version doesn't get a lot of use- only a few people noticed there was no keyboard support in v3.
https://github.com/andykorth/cocos2d-iphone/blob/CCDirector/MergeSchedulerAndActionManager/cocos2d/CCResponderManager.m#L549

rkachowski replied:
wrt the discussion: in what situation would you wish to have more than one CCView in an application? The concept of multiple directors makes sense in this situation if you have more than one view and wish to display multiple scenes simultaneously, but I can't seem to envision a use case that can't be achieved with render targets.

As far as testability goes, I feel this issue is due to the dependencies of discrete objects within cocos2d on a director rather than the fact that it exists as a singleton. I'd like to point out that I'm in no way attached to the implementation of the director as a singleton.

I'm totally for moving the platform specific code out of the director instances, this would really help differentiate between platform / GL initialization setup and what is relevant for cocos configuration. 

I've started with a pull request to make this clearer with regards to https://github.com/spritebuilder/SpriteBuilder/issues/912, but it has significant crossover with your changes here and I don't want to complicate your efforts (esp with the impending v4 deadline) so I will wait for this to go in.

andykorth committed:
Remove a layer of indirection with mac NSResponder methods. Do the director binding in CCResponderManager.

andykorth commented:
Thoughts on these changes? Removed all the redirection methods from CCRespondermanager here: https://github.com/andykorth/cocos2d-iphone/commit/45d0184a8e3a5db6d8533d8ccc4db29aa7fb3b97#diff-9359b51e5d023b91282b75346faf53e6L607

Made the CCViewMacGL ones shorter, and call the right methods directly. 
https://github.com/andykorth/cocos2d-iphone/commit/45d0184a8e3a5db6d8533d8ccc4db29aa7fb3b97#diff-df9801ff9df3d2733cc1388166fd851cL175

Moved the bindDirector calls into the actual responder methods, the main plus being that there's fewer of them there. I think the event handling methods in CCResponderManager could be integrated a bit more, but that will have to wait til after the Global Game Jam!

NickyWeber commented:
Is there actually demand to make games with more than one view?

rkachowski commented on cocos2d/CCActionManager_Private.h:
```diff
@@ -24,5 +24,5 @@
 
 @interface CCActionManager ()
--(void)migrateActions:(id)target from:(CCActionManager*)oldManager;
+// TODO eliminate file
```
why not remove it now?

rkachowski commented on cocos2d/CCDirector.h:
```diff
@@ -66,13 +66,13 @@
  If you want to pause your animation call [pause] instead.
  */
--(void) stopAnimation;
+-(void) stopRunLoop;
```
The comment above will need to be changed to reflect the functionality and so the documentation doesn't conflate "animation" and "runloop"

rkachowski commented on cocos2d/CCDirector.h:
```diff
@@ -66,13 +66,13 @@
+ @warning Don't call this function to start the main loop. To run the main loop call presentScene
  */
--(void) startAnimation;
+-(void) startRunLoop;
```
I would recommend renaming this, as having a method `startRunLoop` along with the attached warning that states "don't call this to start the loop" is pretty counter intuitive

andykorth replied:
Ah. I think that warning means the user should never call this method. Honestly it should be in a private header. 

It is the method that starts the main loop, but the user is supposed to present a scene, which internally calls this.

The only situations when the user may want to start and stop the run loop is- when a mobile app is suspended (as a lifecycle event), or if they are moving in and out of a cocos2d game within something like a UIKit-based app. 

"startAnimation" was a very confusing name, I thought it had something to do with CCAnimations, etc. I'm open to other suggestions, other than "startRunLoop". But is this problem fixed if I just correct the comments and move these to a private header?

rkachowski replied:
Yeah a private header would hide this from a user and fixing the comment would make more sense.

rkachowski commented on cocos2d/CCDirector.h:
```diff
@@ -489,15 +472,15 @@ typedef NS_ENUM(NSUInteger, CCDirectorProjection) {
 
 /** Begins drawing the screen. Scheduled timers and actions will run.
  
  @warning Don't call this function to start the main loop. To run the main loop call presentScene:
```
as above

rkachowski commented on cocos2d/CCDirector.m:
```diff
@@ -160,30 +153,10 @@ - (id) init
 		_displayStats = NO;
 		_totalFrames = _frames = 0;
-
-		// paused ?
```
removing these kind of comments is awesome :+1:

rkachowski commented on cocos2d/CCLabelTTF.m:
```diff
@@ -479,7 +477,7 @@ - (CCTexture*) createTextureWithAttributedString:(NSAttributedString*)attributed
     CGSize originalDimensions = _dimensions;
   
-    CGFloat scale = [CCDirector sharedDirector].contentScaleFactor;
+    CGFloat scale = __ccContentScaleFactor;
```
i would change this to  `CGFloat scale = self.director.contentScaleFactor` as you have on line 706

rkachowski commented on cocos2d/CCNode_Private.h:
```diff
@@ -81,4 +75,22 @@ CGPointApplyGLKMatrix4(CGPoint p, GLKMatrix4 m){
+ 
+ @since v4.0
+ */
+-(void)setRawParent:(CCNode *)parent;
```
~~The comment quite clearly explains the use for this, but it feels a bit awkward to add this property to every instance + subclass of `CCNode`, just to support the stencil / clipping node functionality.~~

forget that, i just noticed this is a private header :)

rkachowski commented on cocos2d/CCResponderManager.m:
```diff
@@ -695,29 +629,49 @@ - (void)scrollWheel:(NSEvent *)theEvent
-    
-}
+    if (!_enabled) return;
+    //  [[CCDirector currentDirector].view.window makeFirstResponder:[CCDirector currentDirector].view];
```
remove?

rkachowski commented on cocos2d/CCResponderManager.m:
```diff
@@ -695,29 +629,49 @@ - (void)scrollWheel:(NSEvent *)theEvent
+- (void)keyUp:(NSEvent *)theEvent
 {
+    if (!_enabled) return;
+    //  [[CCDirector currentDirector].view.window makeFirstResponder:[CCDirector currentDirector].view];
```
also

rkachowski commented on cocos2d/CCScene+Private.h:
```diff
@@ -1,9 +1,7 @@
 /*
```
git has really screwed up and thinks this is a rename operation..

andykorth replied:
Speaking of which, CCScene+Private.h should be CCScene_Private.h to be consistent. Will fix.

Apparently the only common lines is the header and the "@end". Haha

rkachowski commented on cocos2d/CCScheduler.m:
```diff
@@ -71,32 +76,190 @@ @interface CCTimer (Private)
+// This is for performance reasons so that we don't need to copy large arrays every frame to iterate them safely.
+// Instead, the copy is only performed when the array is actually changed.
+
+@interface CopyOnWriteArray : NSObject<NSFastEnumeration> @end
```
would you consider moving this into its own file?

andykorth replied:
Hmm, maybe. It's a completely internal class though, only used within CCScheduler.m

Just as a matter of length of the file, CCTimer could be moved out. It is used externally. (and in user code)

rkachowski commented on cocos2d/CCScheduler.m:
```diff
@@ -137,19 +300,20 @@ -(void)setPaused:(BOOL)paused
 @end
 
 
+//MARK: CCTimer
```
is this some kind of IDE specific instruction, or do you mean `#pragma mark` ?

andykorth replied:
In xcode, `#pragma mark` and `//MARK` do the same thing. Consistency would be good. There's 692 `#pragma mark`s and 120 `//MARK:`s. I'll change the comments to pragmas for consistency, and I'll skip the ones in chipmunk (which looks like most of them)

slembcke replied:
Actually I'd do it the other way around. The pragma version came first, then the comment version. I think the reason was because pragmas are supposed to be for compilers, not IDEs. You are actually supposed to put ifdefs around pragmas since they are compiler specific and will cause an errors in other compilers. (Not that it's a problem here)

andykorth replied:
Hm, I initially agreed, but if I'm going to find/replace them all, that should be a separate pull request. I don't want to delay this or make the pull request even bigger.

rkachowski commented on cocos2d/CCScheduler.m:
```diff
@@ -517,7 +693,107 @@ -(void)update:(CCTime)dt
 	
 	InvokeMethods(_updates, @selector(update:), clampedDelta);
+    
+    #warning TODO: also invoke update actions:
```
the line below invokes `updateActions` or does this warning refer to something else?

andykorth replied:
This is one of the unfinished items. Specifically fixed update actions.

rkachowski commented on cocos2d/CCTextureCache.m:
```diff
@@ -82,9 +82,13 @@ -(id) init
+
+#if __CC_PLATFORM_IOS
+		CCViewiOSGL *view = (CCViewiOSGL*)[[CCDirector currentDirector] view];
 		NSAssert(view, @"Do not initialize the TextureCache before the Director");
```
you can move this comment outside of the `#if` clause

rkachowski commented on cocos2d/CCTextureCache.m:
```diff
@@ -82,9 +82,13 @@ -(id) init
 		NSAssert(view, @"Do not initialize the TextureCache before the Director");
+#elif __CC_PLATFORM_MAC
+        CCViewMacGL *view = (CCViewMacGL*)[[CCDirector currentDirector] view];
+        NSAssert(view, @"Do not initialize the TextureCache before the Director");
```
same

rkachowski commented on cocos2d/CCTextureCache.m:
```diff
@@ -238,7 +242,7 @@ -(void) addImageAsync: (NSString*)path target:(id)target selector:(SEL)selector
 //            [EAGLContext setCurrentContext:nil];
```
i would gratuitously kill all these comments instead of just updating the director referencing - the previous code exists in version control should we need to reference it

rkachowski commented on cocos2d/Platforms/Android/CCGLView.h:
```diff
@@ -41,9 +41,8 @@ enum CCAndroidScreenMode {
-BRIDGE_CLASS("com.apportable.GLView")
-@interface CCGLView : GLView <CCDirectorView>
-
+BRIDGE_CLASS("org.cocos2d.CCGLView")
```
is this correct? afaik there should be no `.java` files in cocos2d, and instead bridged from `GLActivityKit` - there is no `CCGLView.java` class to bridge with

slembcke replied:
Not sure if those changes were merged into this branch. (There are too many to keep track of right now...)

andykorth replied:
I'm going to CC @osinoleg on this. I haven't looked into any of his android changes here. I know he's working on it.

rkachowski commented on cocos2d/Platforms/Mac/CCViewMacGL.m:
```diff
@@ -0,0 +1,232 @@
+
+- (NSUInteger) depthFormat
+{
+    return 24;
```
i would create a named constant to give context to this number (i.e. bitdepth)

rkachowski commented on cocos2d/Platforms/Mac/CCViewMacGL.m:
```diff
@@ -0,0 +1,232 @@
+    if(director.runningScene){
+        [director mainLoopBody];
+    }
+    //	[self setNeedsDisplay:YES];
```
remove?

rkachowski commented:
My main concern is that the `currentDirector` calls rely on an implicit state existing that isn't directly clear to the end user, where operations will tend to fail if bind/unbind isn't called around calls when the user attempts to interact with the director.

andykorth commented:
@NickyWeber Good question. While in theory it can be done, I think real benefits occur in situations like: UIKit apps that run some cocos2d content inside a mostly Cocoa app, which they might navigate away from and back to. Support for multiple views is mostly just a side benefit of un-singletoning the CCDirector.

slembcke commented:
Right. It's actually quite difficult, and _extremely_ poorly documented on how to use Cocos2D within a larger application, even if you only want one Cocos view at a time. This does come up pretty often.

One of the benefits of SpriteKit is that it works in a regular UIKit/AppKit way. A SKView is just another view that you can create and destroy as you please. I can't count the number of times I've read comments about how this is a killer SK feature. Cocos2D was designed to control the entire app, and doesn't like to share.

Think about the way Cocos2D integrates with SB. You have a single document view that is open for the entire duration of the app. Lots of shared state when you close one project and open another. No possible support for having two projects open at once. SB is designed around these limitations.

@rkachowski 
We should be able to cover all of the entry points pretty easily for anybody using Cocos in a traditional way (one director/view for the entire app's lifecycle). We'll need to document how currentDirector works for the people that want to do more, but it's not more complicated than other contexts (GL, CoreAnimation, etc). I agree that contexts are annoying, and prefer APIs without them, but I don't have an alternative that makes it possible to extend the functionality and still remain backwards compatible.

andykorth committed:
Code review cleanup

andykorth committed:
Merge branch 'develop' of github.com:cocos2d/cocos2d-swift into CCDirector/MergeSchedulerAndActionManager

andykorth committed:
Clarifying CCDirectorDelegate and its purpose

andykorth committed:
Fix iOS CCDirector references.

andykorth committed:
Fix iOS startup.

andykorth committed:
Merge branch 'develop' of github.com:cocos2d/cocos2d-swift into CCDirector/MergeSchedulerAndActionManager

Conflicts:
	cocos2d-tests.xcodeproj/project.pbxproj
	cocos2d.xcodeproj/project.pbxproj
